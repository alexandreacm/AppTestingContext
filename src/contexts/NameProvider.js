import React, { createContext, useContext, useState } from "react";

const NameContext = createContext({});

export const useNameContext = () => useContext(NameContext);

function NameProvider({ children }) {
  const [name, setName] = useState(null);
  const [error, setError] = useState(null);

  function saveName(name) {
    console.log(name);
    console.log(error);

    if (name !== null) {
      setName(name);
      setError(null);
    } else {
      setError("Please, insert valid name");
    }
  }

  return (
    <NameContext.Provider value={{ name, setName, setError, error }}>
      {children}
    </NameContext.Provider>
  );
}

export default NameProvider;
