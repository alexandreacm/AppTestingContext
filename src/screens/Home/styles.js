import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  input: {
    width: 350,
    borderWidth: 1,
    borderColor: "#CDCDCD",
    borderRadius: 10,
    padding: 10,
  },
  headerText: {
    fontSize: 35,
    fontWeight: "bold",
    textAlign: "center",
    marginBottom: 10,
    color: "#353031",
  },
  error: {
    fontSize: 30,
    fontWeight: "bold",
    textAlign: "center",
    marginBottom: 10,
    color: "#FF0000",
  },
  buttonText: {
    fontSize: 15,
    fontWeight: "bold",
    textAlign: "center",
    color: "#FFF",
  },
  button: {
    width: 350,
    height: 50,
    backgroundColor: "#993399",
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 8,
    marginTop: 15,
  },
});
