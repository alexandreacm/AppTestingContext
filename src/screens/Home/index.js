import React, { useState, useRef } from "react";
import {
  Text,
  TextInput,
  TouchableOpacity,
  KeyboardAvoidingView,
} from "react-native";

import { useNameContext } from "../../contexts/NameProvider";
import { styles } from "./styles";

export function Home({ name, setName, setError, error }) {
  const [newName, setNewName] = useState(null);
  const ref = useRef(0);

  function onHandleSaveName() {
    if (newName === null) {
      setError("Please, insert valid name");
    } else {
      setName(newName);
      setNewName(null);
      setError(null);
    }
  }

  return (
    <KeyboardAvoidingView behavior="position">
      {error ? (
        <Text testID="error" style={styles.error}>
          {error}
        </Text>
      ) : (
        <Text testID="text" style={styles.headerText}>
          Welcome, {name || "No Name"}
        </Text>
      )}

      <TextInput
        style={styles.input}
        testID="name"
        label="Name"
        value={newName}
        onChangeText={setNewName}
        placeholder="Please, typing here...."
      />

      <TouchableOpacity
        testID="buttonChange"
        style={styles.button}
        onPress={onHandleSaveName}
      >
        <Text style={styles.buttonText}>Change</Text>
      </TouchableOpacity>
      <Text>Renders: ..{ref.current++}</Text>
    </KeyboardAvoidingView>
  );
}

export default (props) => {
  const { name, setName, saveName, setError, error } = useNameContext();
  return (
    <Home
      name={name}
      setName={setName}
      saveName={saveName}
      setError={setError}
      error={error}
      {...props}
    />
  );
};
