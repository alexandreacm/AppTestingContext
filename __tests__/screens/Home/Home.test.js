import { create } from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react-native";

import Home from "../../../src/screens/Home";
import NameProvider from "../../../src/contexts/NameProvider";

describe("All tests", () => {
  const tree = create(<Home />).toJSON();

  test("Should render correctly", () => {
    create(<Home />);
  });

  test("Should render correctly Home", () => {
    render(<Home />);
  });

  test("Should render correctly snapshot", () => {
    expect(tree).toMatchSnapshot();
  });

  test("Should display default text No Name", () => {
    const { getByText } = render(<Home />);

    const textNoName = getByText("Welcome, No Name");
    expect(textNoName).not.toBeNull();
    //expect(textNoName).not.toBe(null);
  });

  test("Should display text Welcome, alex", () => {
    const { getByText, getByTestId } = render(<Home />, {
      wrapper: NameProvider,
    });

    const input = getByTestId("name");
    fireEvent.changeText(input, "alex");

    const button = getByTestId("buttonChange");
    fireEvent.press(button);

    const textName = getByText("Welcome, alex");
    expect(textName).not.toBeNull();
    //expect(textNoName).not.toBe(null);
    //expect(textNoName.length).toBe(1);
  });

  test("Should not display default text No Name", () => {
    const { queryByText, getByTestId } = render(
      <NameProvider>
        <Home />
      </NameProvider>
    );

    const input = getByTestId("name");
    fireEvent.changeText(input, "Alex");

    const button = getByTestId("buttonChange");
    fireEvent.press(button);

    expect(queryByText("Welcome, No Name")).toBeNull();
    //expect(queryByText("Welcome, No Name")).toBe(null);
  });

  test("Should display a text with Welcome, alex", () => {
    const { getByText } = render(<Home name="Alex" />);

    getByText("Welcome, Alex");
  });

  test("Should call setName function", () => {
    const setName = jest.fn();
    const setError = jest.fn();

    const { getByTestId } = render(
      <Home setName={setName} setError={setError} />
    );

    const input = getByTestId("name");
    fireEvent.changeText(input, "alex");

    const button = getByTestId("buttonChange");
    fireEvent.press(button);

    expect(setName).toBeCalledWith("alex");
  });
});
