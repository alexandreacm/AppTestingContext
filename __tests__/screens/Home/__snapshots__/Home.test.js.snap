// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All tests Should render correctly snapshot 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      Object {
        "bottom": 0,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#353031",
          "fontSize": 35,
          "fontWeight": "bold",
          "marginBottom": 10,
          "textAlign": "center",
        }
      }
      testID="text"
    >
      Welcome, 
      No Name
    </Text>
    <TextInput
      allowFontScaling={true}
      label="Name"
      onChangeText={[Function]}
      placeholder="Please, typing here...."
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "#CDCDCD",
          "borderRadius": 10,
          "borderWidth": 1,
          "padding": 10,
          "width": 350,
        }
      }
      testID="name"
      underlineColorAndroid="transparent"
      value={null}
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#993399",
          "borderRadius": 8,
          "height": 50,
          "justifyContent": "center",
          "marginTop": 15,
          "opacity": 1,
          "width": 350,
        }
      }
      testID="buttonChange"
    >
      <Text
        style={
          Object {
            "color": "#FFF",
            "fontSize": 15,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Change
      </Text>
    </View>
    <Text>
      Renders: ..
      0
    </Text>
  </View>
</View>
`;
